clustering/ejb/WFLY-7628_Distributed_EJB_Timers.adoc

```diff
@@ -56,87 +59,108 @@ WildFly needs an persistent timer service implementation suitable for use in a c
 
-* Servers started using an old version of the EJB subsystem configuration will continue to behave as normal, using the existing TimerServiceImpl for both transient and persistent timers.
+* Servers started using the default ejb3 subsystem configuration will continue to behave as normal, using the existing in-memory `TimerService` implementation for both transient and persistent timers.
 
 === Default Configuration
 
-The following caches will be added to the default ejb cache-container configuration with the Infinispan subsystem:
+The following caches will be added to the default `ejb` cache-container configuration with the Infinispan subsystem:
 
 For the HA profiles:
 
-	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent-timers:add()
-	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent-timers/component=locking:add(isolation=PESSIMISTIC)
-	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent-timers/component=transaction:add(mode=BATCH)
-	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent-timers/store=file:add(purge=false)
+	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent:add()
+	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent/component=locking:add(isolation=PESSIMISTIC)
+	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent/component=transaction:add(mode=BATCH)
+	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent/store=file:add(purge=false)
 
 And for non-HA profiles:
 
-	/subsystem=infinispan/cache-container=ejb/local-cache=persistent-timers:add()
-	/subsystem=infinispan/cache-container=ejb/local-cache=persistent-timers:component=locking:add(isolation=PESSIMISTIC)
-	/subsystem=infinispan/cache-container=ejb/local-cache=persistent-timers:component=transaction:add(mode=BATCH)
-	/subsystem=infinispan/cache-container=ejb/local-cache=persistent-timers/store=file:add(purge=false,passivation=false)
+	/subsystem=infinispan/cache-container=ejb/local-cache=persistent:add()
+	/subsystem=infinispan/cache-container=ejb/local-cache=persistent:component=locking:add(isolation=PESSIMISTIC)
+	/subsystem=infinispan/cache-container=ejb/local-cache=persistent:component=transaction:add(mode=BATCH)
+	/subsystem=infinispan/cache-container=ejb/local-cache=persistent/store=file:add(purge=false, passivation=false)
 
 The following timer management provider will be added to the default distributable-ejb subsystem configuration:
 
-	/subsystem=distributable-ejb/infinispan-timer-management=persistent:add(cache-container=ejb, cache=persistent-timers)
+	/subsystem=distributable-ejb/infinispan-timer-management=persistent:add(cache-container=ejb, cache=persistent)
 
-The default timer-service configuration of the EJB subsystem will be modified as follows:
+To enable use of the distributable timer service, a user will be required to modify the default timer-service configuration of the ejb3 subsystem as follows:
```

Comment:
Looking at the following line below: 

`/subsystem=ejb3/service=timer-service:undefine-attribute(name=default-data-store)`

previously it is stated as a requirement that this attribute should be made optional, so why would the user be requested to unset it?

Reply:
Because this attribute is defined in the default ejb3 subsystem configuration.

Reply:
Ok, thanks for your feedback.

So, since the implementation of this feature will make the attribute optional in the model, the user will be able to _undefine_ it with the proposed command.
I am thinking at users which will use the feature in future versions of WildFly and who never had any kind of previous experience - neither they would need it - with database based persistence. 
Isn't this approach too _indirect_ in such a use case (and maybe in general)? Do you think there can be another solution for enabling the feature without relying on the semantics of a completely different implementation?

Reply:
It's not ideal - but it was done this way in anticipation of the end result.  However, given that the scope of the distributable timer service implementation encompasses the file and database persistence use cases, the expectation here is that this new implementation will eventually replace the existing ones (meaning, that default-data-store, as well as the data store resources themselves, will be deprecated, and the default configuration will no longer define a default-data-store attribute).  At that point, this will no longer be confusing.

Reply:
In which way it will no longer be confusing? Because the attribute will disappear and the users won't have to enable the distributed EJB timers management by executing this command?

In such a case IMHO it is not ideal, as you stated as well, and should be changed, since the proposal is still being worked on, and there will be users (and customers when it comes to the productized version of this feature) relying on its configuration, which shouldn't be "non-ideal" since the beginning.

Isn't there a more straightforward way to enable this?

Reply:
> In which way it will no longer be confusing?

Because timer management will use the distributable implementation by default, i.e. exclusively configured via the distributable-ejb subsystem in our default configuration.

> Because the attribute will disappear

It does not "disappear" - it would be undefined by default.

> Isn't there a more straightforward way to enable the feature?

Such as?  The problem is not just "how to enable the feature", but how to disable the existing configuration.  Ideally, the user performs a single action for both.  Here's a the 3 most obvious options:

1. My first thought was to create a separate distributable-timer-service resource that is mutually exclusive with the existing timer-service resource, which would normally be the cleanest way to introduce a new implementation mean to replace an existing implementation.
     * This prohibits use of different timer service implementations for transient vs persistent timers (not ideal)
     * This suggests requiring a separate distributable-timer-service deployment descriptor namespace used to override distributable timer service behavior per EJB (i.e. more awkward)
     * This complicates any future plans to deprecate features of the existing timer-service, where persistent vs transient timers are not implemented independently.
     * For the above reasons, I discarded this idea
1. The second option is to add separate `default-persistent-timer-management` and `default-transient-timer-management` attributes to the existing timer-service resource, which reference named timer-management providers as defined in the distributable-ejb subsystem.  The `default-persistent-timer-management` attribute would be mutually exclusive with the "default-data-store` attribute.
     * This is more flexible than the proposal 1, since it lets users configure different timer service implementations for transient vs persistent timers
     * From a user perspective, this is more cumbersome than what is described in the current draft of the analysis document.  Enabling this feature now becomes a matter of undefining the `default-data-store` attribute and defining the `default-persistent-timer-management` attribute within a batch (since one of the pair of attributes must be defined).
1. The third option, and what is described in the current draft of the analysis document, is the same as proposal 2, except that the default-*-timer-management attributes reside in the distributable-ejb subsystem instead of the timer-service resource of the ejb3 subsystem.
     * This means that enabling this feature is simply a matter of undefining the `default-data-store` attribute (since the`default-persistent-timer-management` will already be defined in the distributable-ejb subsystem.
     * Additionally, this option was preferred by @rachmatowicz

@fabiobrz What are your thoughts re: proposal 2 vs 3?
@chengfang Any opinion?

Reply:
After some follow-up discussions with team members, option 2 is going to be problematic for galleon provisioning, which does not handle well the case where a single attribute requires different capabilities depending on whether or not it is defined.  Consequently, I think option 2 is the way to go.

Reply:
+1, I see this is consolidated now.

Reply:
+1, I see this is consolidated now.